backend/sessions.md

```diff
@@ -0,0 +1,129 @@
+2. We retrieve the most recent token of the session. If the token used to call the refresh is not the most recent one,
+   we return the most recent token with its expiration time.
+3. (From now on, we're working with the most recent token). If the token has been issued less than 5 minutes ago,
+   we return the token with its expiration time.
+4. We get a lock for the session.
+5. We call the login-module to refresh the token, with the refresh token of the session, and get back:
+  - the new access token
+  - the new refresh token
+  - the expiration time of the access token
+6. We store the refresh token in the session, and store the new access token.
+7. We delete expired old tokens for the session.
+
+
+### Only the most recent token can be used to refresh the token
+
+If another token is used to refresh the token, the current most recent token is returned with its expiration time.
+
+
+### A token can only be refreshed once
+
+Because a token can only be refreshed by the most recent token, it implies that a token can only be refreshed once.
+
+
+### A token can only be refreshed if it's not expired
+
+If an expired token is used to refresh the token, access defined error is returned.
+
+
+### A token can only be refreshing 5 minutes after it was issued
+
+It's not possible to refresh a token that was issued less than 5 minutes ago.
+
+This was done because:
+- There can be many tabs open in the browser. Each tab will ask for a refresh of the token at roughly the same time.
+  We don't want to have many refreshes going at the same time: it's completely useless.
+
+
+### There a maximum number of access tokens for a session that can be present in database at one moment
+
+This follows from the facts that:
+- a token can only be refreshed once
+- a token can only be refreshed after 5 minutes
+- expired tokens are deleted on refresh
+
+Since tokens are valid for 2 hours, we can have a maximum of 2 hours / 5 minutes = 24 tokens in the database at the same time.
+
+
+## Logout
+
+When a user logs out, the session and all the related tokens are deleted.
+
+
+## Edge cases with the frontend
+
+The tricky edge cases come from the frontend when:
+- tabs are frozen when inactive and in background for some time
+- the device goes into sleep
+
+
+### What happens if the frontend tries to refresh a token that was already refreshed (that is not the most recent one)?
+
+It will receive the most recent token with its expiration time.
+
+With the expiration time, it will know when the token has to be refreshed again. Which can be right now.
+
+
+### What happens when many tabs make a refresh request at the same time?
+
+There is a lock for the session for the refresh process.
+
+1. The first request gets the lock, and the token gets refreshed.
+2. The other requests (already made before the refresh was done), are using tokens that are not the most recent anymore.
+   The most recent token (just refreshed) will be sent with its expiration time.
```

Comment:
Not clear, it will not be sent to the other requests, so it is not related with (2).